fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/FallbackCursor.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * FallbackCursor.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2022 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.cursors;
+
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.record.RecordCursor;
+import com.apple.foundationdb.record.RecordCursorResult;
+import com.apple.foundationdb.record.RecordCursorVisitor;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.function.Supplier;
+
+/**
+ * provide an alternative cursor in case the primary cursor fails. This cursor has an <code>inner</code> cursor that is
+ * used to return the results in the sunny day scenarios, and an alternative provider of a <code>fallback</code> cursor
+ * that will be used if the primary cursor encounters an error.
+ * Note that there are business rules around when a fallback may happen. For example, because of the way records are
+ * returned to the consumer, once a record is returned from the primary cursor, no fallback may be permitted: This is
+ * done in order to prevent the case where a few records are returned, then a failure happens and the fallback cursor
+ * starts again from the beginning, resulting in duplicate records being returned.
+ * In practice, since many errors are observed when the request is sent to FDB (which coincide with the cursor's
+ * first <code>onNext()</code> call, many such failures will be caught by that first result.
+ *
+ * @param <T> the type of cursor result returned by the cursor
+ */
+@API(API.Status.MAINTAINED)
+public class FallbackCursor<T> implements RecordCursor<T> {
+    @Nonnull
+    private final Supplier<RecordCursor<T>> fallbackCursorSupplier;
+    @Nullable
+    private final Executor executor;
+    @Nonnull
+    private RecordCursor<T> inner;
+    @Nullable
+    private RecordCursorResult<T> nextResult;
+    @Nullable
+    private CompletableFuture<RecordCursorResult<T>> nextResultFuture;
+
+    private boolean alreadyFailed = false;
+    private boolean allowedToFail = true;
+
+    public FallbackCursor(@Nonnull RecordCursor<T> inner, @Nonnull Supplier<RecordCursor<T>> fallbackCursorSupplier) {
```

Comment:
Regarding continuations, I think our general approach would be to wrap the continuations in some kind of protobuf message, and then to include information about runtime state in the continuation. For example, that's what the `OrElse` cursor does:

https://github.com/FoundationDB/fdb-record-layer/blob/c7b9066ebfae3972422e6ec0a8710e3dd4f24974/fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/OrElseCursor.java#L131

That being said, for the particular case of index prefetch, it's pretty compelling that the continuations from both the traditional index scan and index prefetch paths are identical, which means that the choice can be purely an execution detail. We should probably at least document that the main cursor and fallback cursor need to have compatible continuations.

Reply:
Agree. Created issue to wrap the continuation and restore the state. One note here is that the continuations are assumed to have been given to the `inner` and `fallback` cursors respectively before the `FallbackCursor` was created, so that would work even if the continuations would not match. The thing we are missing here is the ability to recreate the state of the fallback cursor (i.e. continue from the "failed" state).